### PR TITLE
folder level wild card support for COPY FROM S3 and file scheme

### LINKED ISSUE #10915: Folder wildcard support for `COPY FROM` s3
**Use case**:

As a CrateDB user I want to copy data from all files in multiple folders in S3 with a single `COPY FROM` command into CrateDB.

**Feature description**:

> We have the following s3 folders structure with files: bucket/year=2020/month=[1-12]/day=[1-31]/hour=[1-24]/file[N].gz.
> I want to copy all files for example from bucket/year=2020/month=12/ folder using one command

Wildcard works for filenames: ```COPY target FROM 's3://{aa:bb}@bucket/year=2020/month=12/day=10/hour=12/*.json```

It would be great if CrateDB could support wildcards for folder names as well.

`COPY FROM` with `file` already supports `*` wildcard for folder names (but this is undocumented behaviour):
```COPY target FROM '/temp/*/*.json```

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB

`HAS_GLOB_PATTERN` seems to have a bug that caused the folder level wildcard expansion for `file://` scheme acted like it was supported. It does not make sense to suddenly disable it for file scheme, so I am trying to support folder level wildcard expansion for both s3 and file scheme. 

CrateDB will now accept for example, `copy t from 's3://bucket/year=2020/month=*/day=*/hour=*/file*'`. It will match URIs like `s3://bucket/year=2020/month=1/day=12/hour=3/file100.json`. 

closes #10915 

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
other URL protocols do get impacted but it either
1) finds the resource specified by the URL without wildcard or
2) fails to find the resource specified by the URL with wildcard
so no further action required


Integration testing with minio does not seem to work for my purpose since what I need to test is `copy t from 's3://bucket/year=2020/month=12/day=10/hour=12/*.json'` 
Initially, the issue was that a wildcard still remained in the prefix that was passed to `AmazonS3#listObjects`, so instead of itests, I mocked it so that I can verify the arguments passed. It seems sufficient but I would like to do manual testing on it.

**Comment by jeeminso** [comment]
PR checks passed previously but failed after a minor doc change. will look into it tomorrow.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java [hunk 1]
```diff
@@ -820,4 +821,78 @@ public void test_copy_from_csv_file_with_empty_string_as_null_property() throws
             printedTable(response.rows()),
             is("1| foo\n2| NULL\n3| NULL\n"));
     }
+
+    @Test
+    public void testCopyFromFileSchemeWithFileLevelWildCards() throws Exception {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:826]: Would it be possible to write unit tests against `getUrisWithGlob` directly instead?

Would be a lot less expensive than running the full integration tests.

File: server/src/test/java/io/crate/integrationtests/SwapTableITest.java [hunk 2]
```diff
@@ -63,6 +64,7 @@ public void testSwapTwoTablesWithDropSource() {
                                 4041));
     }
 
+    @Repeat(iterations = 50)
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/SwapTableITest.java:67]: Unrelated?

**Review by mfussenegger — COMMENTED** [review]
Looks good :+1: 

Left a minor comment regarding the testing approach

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 1 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: @mfussenegger Hello Jordi I have made some changes accordingly. Please have another look.

1) I found out that my fix has inconsistency where for file scheme it kept trailing slash while s3 does not. I think it still works but made a fix for that.
2) removed most of the test cases, integration testing for file scheme and mocked s3 test cases. They are then replace by the `getUrisWithGlob` test.
3) in order to test `getUrisWithGlob` relaxed some of the visibilities.

Comment (jeeminso) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: and I should revert the changes of 'S3FileInputTest` instead of just removing my test cases.

File: server/src/test/java/io/crate/integrationtests/SwapTableITest.java (hunk 2 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/integrationtests/SwapTableITest.java:67]: Thanks for catching this. I will take this out..

File: docs/sql/statements/copy-from.rst [hunk 3]
```diff
@@ -114,6 +116,10 @@ URI
 A string literal or array of string literals containing URIs. Each URI must be
 formatted according to the `URI Scheme`_.
 
+The ``COPY FROM`` statement allows ``URIs`` with wildcards to conveniently match all
```
Comment (jeeminso) [inline docs/sql/statements/copy-from.rst:119]: @norosa Hello Nomi, I did a little bit of doc refactoring since now URIs with folder level wildcard is allowed for both S3 and file schemes. I would appreciate if you could have quick look. Thanks!

File: docs/sql/statements/copy-from.rst [hunk 4]
```diff
@@ -114,6 +116,10 @@ URI
...
+directories and files. So, ``file:///tmp/import_data/*/*.json`` would expand to include
+every files with JSON extension within the immediate sub-directories of ``/tmp/import_data``.
+
```
Comment (norosa) [inline docs/sql/statements/copy-from.rst:122]:
```suggestion
You can use filename `globbing`_ (i.e., ``*`` wildcards) with the ``COPY
FROM`` statement to construct URIs that can match multiple directories and
files.

Suppose you used ``file:///tmp/import_data/*/*.json`` as the URI. This URI
would match all JSON files located in subdirectories of the
``/tmp/import_data`` directory.

So, for example, these files would match:

- ``/tmp/import_data/foo/1.json``
- ``/tmp/import_data/bar/2.json``
- ``/tmp/import_data/1/boz.json``

.. CAUTION::

    A file named ``/tmp/import_data/foo/.json`` would also match the
    ``file:///tmp/import_data/*/*.json`` URI. The ``*`` wildcard matches any
    number of characters, including none.

However, these files would match:

- ``/tmp/import_data/1.json`` (two few subdirectories)
- ``/tmp/import_data/foo/bar/2.json`` (too many subdirectories)
- ``/tmp/import_data/1/boz.js`` (incorrect file extension)

.. _globbing: https://en.wikipedia.org/wiki/Glob_(programming)
```

how about this?

(I would move the `globbing` link definition to the bottom of the file

**Review by norosa — COMMENTED** [review]
thx! left a comment for your consideration!

File: docs/sql/statements/copy-from.rst (hunk 4 above)
Comment (jeeminso) [reply docs/sql/statements/copy-from.rst:122]: Thanks @norosa, I have changed a few shown bolded below.
my understanding is that filename is /tmp/**a.json** and pathname is **/tmp/a.json**
Also wildcards include * and few other things, and I think we should clearly say only * is accepted.


You can use **`globbed`_ pathnames (specifically,** ``*`` wildcards) with the ``COPY
FROM`` statement to construct URIs that can match multiple directories and
files.
Suppose you used ``file:///tmp/import_data/*/*.json`` as the URI. This URI
would match all JSON files located in subdirectories of the
``/tmp/import_data`` directory.
So, for example, these files would match:

- ``/tmp/import_data/foo/1.json``
- ``/tmp/import_data/bar/2.json``
- ``/tmp/import_data/1/boz.json``
.. CAUTION::
    A file named ``/tmp/import_data/foo/.json`` would also match the
    ``file:///tmp/import_data/*/*.json`` URI. The ``*`` wildcard matches any
    number of characters, including none.
However, these files would not match:

- ``/tmp/import_data/1.json`` (two few subdirectories)
- ``/tmp/import_data/foo/bar/2.json`` (too many subdirectories)
- ``/tmp/import_data/1/boz.js`` (**file extension mismatch**)

**Review by mfussenegger — APPROVED** [review]

**Comment by jeeminso** [comment]
performed manual testing, s3 with folder level wildcard behaves as expected.


```SQL
bucket/
   tt.json (1234)
   test/
      3.json (3)
      test2/
         4.json (4)
         test3/
            5.json (5)
      test4/
         1.json (1)
         2.json (2)
      
1)
cr> copy tbl from 's3://bucket/*.json';                                           
COPY OK, 1 row affected  (0.534 sec)
cr> select * from tbl;                                                                                                                                        
+------+
|  int |
+------+
| 1234 |
+------+
SELECT 1 row in set (0.002 sec)

2)
cr> copy tbl from 's3://bucket/*/*.json';                                         
COPY OK, 1 row affected  (0.565 sec)
cr> select * from tbl;                                                                                                                                        
+------+
|  int |
+------+
| 1234 |
|    3 |
+------+
SELECT 2 rows in set (0.001 sec)

3)
cr> copy tbl from 's3://bucket/*/*/*.json';                                       
COPY OK, 3 rows affected  (0.642 sec)
cr> select * from tbl;                                                                                                                                        
+------+
|  int |
+------+
| 1234 |
|    3 |
|    4 |
|    1 |
|    2 |
+------+
SELECT 5 rows in set (0.001 sec)

4)
cr> copy tbl from 's3://bucket/*/*/*/*.json';                                     
COPY OK, 1 row affected  (0.495 sec)
cr> select * from tbl;                                                                                                                                        
+------+
|  int |
+------+
| 1234 |
|    3 |
|    4 |
|    1 |
|    2 |
|    5 |
+------+
SELECT 6 rows in set (0.001 sec)

5)
cr> delete from tbl;                                                                                                                                          
DELETE OK, 6 rows affected  (0.014 sec)
cr> copy tbl from 's3://bucket/test/t*2/*.json';                                  
COPY OK, 1 row affected  (0.466 sec)
cr> select * from tbl;                                                                                                                                        
+-----+
| int |
+-----+
|   4 |
+-----+
SELECT 1 row in set (0.927 sec)

6)
cr> copy tbl from 's3://bucket/test/t*4/*.json';                                  
COPY OK, 2 rows affected  (0.665 sec)
cr> select * from tbl;                                                                                                                                        
+-----+
| int |
+-----+
|   2 |
|   4 |
|   1 |
+-----+
SELECT 3 rows in set (0.750 sec)

7)
cr> copy tbl from 's3://bucket/test/test2/test3/*';                               
COPY OK, 1 row affected  (0.552 sec)
cr> select * from tbl;                                                                                                                                        
+-----+
| int |
+-----+
|   5 |
|   2 |
|   4 |
|   1 |
+-----+
SELECT 4 rows in set (0.843 sec)

8)
cr> copy tbl from 's3://bucket/*/test2/test3/*';                                  
COPY OK, 1 row affected  (0.774 sec)
cr> select * from tbl;                                                                                                                                        
+-----+
| int |
+-----+
|   5 |
|   5 |
|   2 |
|   4 |
|   1 |
+-----+
SELECT 5 rows in set (0.398 sec)

```

File: docs/sql/statements/copy-from.rst (hunk 4 above)
Comment (norosa) [reply docs/sql/statements/copy-from.rst]: "i.e.," means the same as "specifically", so I would keep "i.e.," (you might've been thinking of "e.g.," which means "for example")

you are right about the difference between _filenames_ and _paths_. I would keep the grammar of the opening sentence though, so: "You can use pathname globbing...", because:

- "You can use globbed pathnames..." does not specify who is globbing (passive voice)
- "You can use pathname globbing..." makes it clear that the user can do the globbing (active voice)

"file extension mismatch" is a good change!

Comment (norosa) [reply docs/sql/statements/copy-from.rst]: I just noticed this was already merged. my bad! I am working on a small PR to add some copyedits to this file and will add you as a reviewer @jeeminso

Comment (norosa) [reply docs/sql/statements/copy-from.rst]: copyedits here: https://github.com/crate/crate/pull/11449
